### ISSUE TITLE
Mini.Sessions breaks with nvim-notify

### ISSUE THREAD

**Issue by kvietcong** [echasnovski/mini.nvim#79]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)
- [X] I have updated 'mini.nvim' to latest version

### Module(s)

mini.sessions, mini.starter

### Description

Here is my auto save autocommand for saving the last session automatically:
```lua
vim.api.nvim_create_autocmd("VimLeavePre", {
    callback = function() minisessions.write("LastSession.vim", {}) end
})
```

This works if I leave and come back using mini.starter. However, if I were to use that session and quit out again, coming back breaks it. A new empty buffer appears and every window size is severely out of proportion.

First Time Loading:
![First Time Loading](https://user-images.githubusercontent.com/32186663/167513188-9f72b053-804c-4577-ab9b-9a1090b89ef2.png)

Second Time Loading:
![Second Time Loading](https://user-images.githubusercontent.com/32186663/167513271-aa121db3-6bbd-4664-8232-e196f7b92480.png)

However I did find the problem. When I make nothing verbose in the config, everything works fine. Also, if I disable nvim notify, everything also works. I see in the implementation that notify is used for verbose so I'm assuming it has something to do with nvim-notify doing something weird with the buffers. Is there any way to have the verbosity with nvim notify (so that I could see when I manually save/read)?

### Neovim version

0.7.0

### Steps to reproduce

1. `nvim -nu minimal.lua` (with nvim-notify enabled, replacing vim.notify and verbose enabled for writes)
2. Close with auto write last session
3. Open that session and close.
4. Open session again.

### Expected behavior

Sessions load without new buffers or weird proportion sizes.

### Actual behavior

Sessions are loaded with new buffers and weird proportion sizes.

**Comment by echasnovski** [comment]
First of all, is there a reason for using custom autocommand instead of default `MiniSessions.config.autowrite`? It does [almost exactly that](https://github.com/echasnovski/mini.nvim/blob/ee7c02297cf95cba07f8bf8da3ecf99337a1c15e/lua/mini/sessions.lua#L81).

And yes, this might be due to some friction between 'mini.nvim' and 'nvim-notify' as most of user notifications are done with `vim.notify`. I'll look into this.

**Comment by BenjaminKern** [comment]
I had a similiar issue when using https://github.com/rcarriga/nvim-notify. During a session write, a vim notify notification buffer was popping up. When I was closing neovim during this notification and trying to reload the previously saved session, the buffer layout was garbled up (most probably it was trying to save the vim.notify popup somehow)

**Comment by kvietcong** [comment]
> First of all, is there a reason for using custom autocommand instead of default `MiniSessions.config.autowrite`? It does [almost exactly that](https://github.com/echasnovski/mini.nvim/blob/ee7c02297cf95cba07f8bf8da3ecf99337a1c15e/lua/mini/sessions.lua#L81).

I wanted to have a session that is always what I was working on last. I still use autowrite to write if I'm already in a session (like if I were in a project session, it would save, but it would also save it as the last session I'm working on). This way, I don't have to worry about resuming my work if I forget to save the session when I'm not in a session atm.

Also I've seen that this happens with all plugins that have "special buffer usage". Nvim-Tree, Neogit, and other plugins that provide context aware/special buffers often break upon restoration. But they don't garble up the layout nearly as nvim-notify. I think this is just an inherent flaw with mksession. It would be amazing if there was a way to make it work with nvim-notify but I'm definitely doing swell with just disabling the verbose options and making sure I don't save sessions when a notification shows.

Thanks for your hard work!

**Comment by kvietcong** [comment]
A pretty good working solution for this I found was to have a pre write hook that dismisses all notifications like so:

```lua
local minisessions = require("mini.sessions")
minisessions.setup({
    hooks = {
        pre = {
            write = function()
                vim.notify.dismiss({ silent = true, pending = true })
            end,
        },
    },
    verbose = { read = true, write = true, delete = true, },
})
```

I could probably find some way to close all buffers that are creating problems for me through this hook. Feel free to close this issue if this solution is satisfactory for most people.

Thanks

**Comment by echasnovski** [comment]
> I wanted to have a session that is always what I was working on last.

So like a backup session? Nice.

> A pretty good working solution for this I found was to have a pre write hook that dismisses all notifications like so:

Double nice! It is actually a first time someone used this plugin in a way I didn't anticipate beforehand. Kudos!

To make it work with all "special buffers", take a look at `buftype` option. Usually those special buffers have its value set to `nofile`. So making a custom cleanup function using `vim.api.nvim_list_bufs()` can do the trick.

Won't close issue for a bit to ponder about deciding whether to put this kind of cleanup in plugin or leave as help/wiki entry.

**Comment by kvietcong** [comment]
Thanks for the tip! What I was doing before was quite... not great 🤣 

```lua
local close_bad_buffers = function()
  local buffer_infos = vim.split(vim.api.nvim_exec("buffers!", true), "\n")
  local bad_buffers = { "NvimTree", "Neogit", "packer" }
  for _, buffer_info in pairs(buffer_infos) do
      local buffer_id = vim.split(buffer_info, "%s", { trimempty = true })[1]:match("%d+")
      local _, buffer_name, _ = buffer_info:match('(")(.+)(")')
      for _, bad_buffer in pairs(bad_buffers) do
          if string.find(buffer_name, bad_buffer) then
              pcall(vim.cmd, "bd " .. buffer_id)
              break
          end
      end
  end
end
```

I appreciate it a lot as I'm quite new to diving into Neovim (evidently above). Honestly, your plugins have been such a great way to get acquainted into the Neovim ecosystem. Thank you so much for that.

**Comment by kvietcong** [comment]
Oh and also for anyone reading this issue and is suffering from the same problem, here is my new function which might help

```lua
local close_bad_buffers = function()
    require("zen-mode").close()
    vim.notify.dismiss({ silent = true, pending = true })
    local buffer_numbers = vim.api.nvim_list_bufs()
    for _, buffer_number in pairs(buffer_numbers) do
        local buffer_type = vim.api.nvim_buf_get_option(buffer_number, "buftype")
        if buffer_type == "nofile" then
            vim.api.nvim_buf_delete(buffer_number, { force = true })
        end
    end
end
```

**Comment by echasnovski** [comment]
> I appreciate it a lot as I'm quite new to diving into Neovim (evidently above). Honestly, your plugins have been such a great way to get acquainted into the Neovim ecosystem. Thank you so much for that.

And it seems like you are doing a great job. The recent code looks solid!

**Comment by echasnovski** [comment]
It seems that this was fixed in Neovim 0.8.0 stable release. I can not reproduce the error now (autowrite session on Neovim close with notify window shown and then the sessions can be restored without problems). I think neovim/neovim#18635 might have fixed it.

@kvietcong, would you mind double checking if this is still an issue for you?

**Comment by kvietcong** [comment]
It seems like it's all fixed now. I don't need my the custom hooks any longer. Thanks for the update!
